libgeopmd/src/LevelZero.cpp

```diff
@@ -605,71 +632,153 @@ namespace geopm
+                                 ZES_RAS_ERROR_CAT_NON_COMPUTE_ERRORS, M_ERROR_TYPE_UNCORRECTABLE);
     }
 
-    double LevelZeroImp::ras_display_errcount(unsigned int l0_device_idx,
-                                              int l0_domain, int l0_domain_idx) const
+    double LevelZeroImp::ras_cache_errcount_uncorrectable(unsigned int l0_device_idx,
+                                                          int l0_domain,
+                                                          int l0_domain_idx) const
     {
-        return (ras_status_helper(l0_device_idx, l0_domain, l0_domain_idx))[ZES_RAS_ERROR_CAT_DISPLAY_ERRORS];
+        return ras_status_helper(l0_device_idx, l0_domain, l0_domain_idx,
+                                 ZES_RAS_ERROR_CAT_CACHE_ERRORS, M_ERROR_TYPE_UNCORRECTABLE);
     }
 
-    std::array<uint64_t, ZES_MAX_RAS_ERROR_CATEGORY_COUNT> LevelZeroImp::ras_status_helper(unsigned int l0_device_idx,
-                                                                                           int l0_domain,
-                                                                                           int l0_domain_idx) const
+    double LevelZeroImp::ras_display_errcount_uncorrectable(unsigned int l0_device_idx,
+                                                            int l0_domain,
+                                                            int l0_domain_idx) const
+    {
+        return ras_status_helper(l0_device_idx, l0_domain, l0_domain_idx,
+                                 ZES_RAS_ERROR_CAT_DISPLAY_ERRORS, M_ERROR_TYPE_UNCORRECTABLE);
+    }
+
+
+
+    // RAS Helper function that extracts the errorset counters using the cached errorset handle
+    uint64_t LevelZeroImp::ras_status_helper(unsigned int l0_device_idx,
+                                             int l0_domain,
+                                             int l0_domain_idx,
+                                             zes_ras_error_cat_t errorcat,
+                                             int errortype) const
     {
         zes_ras_state_t pState;
-        zes_ras_handle_t handle = m_devices.at(l0_device_idx).
-                                      subdevice.ras_domain.at(l0_domain).at(l0_domain_idx);
+        int ras_idx = (l0_domain_idx * M_NUM_ERROR_TYPE) + errortype;
+        zes_ras_handle_t handle = m_devices.at(l0_device_idx).subdevice.ras_domain.at(ras_idx);
         check_ze_result(zesRasGetState(handle, 0, &pState),
                         GEOPM_ERROR_RUNTIME,
                         "LevelZero::" + std::string(__func__) +
                         ": Sysman failed to get RAS counters",
                         __LINE__);
-
-        std::array<uint64_t, ZES_MAX_RAS_ERROR_CATEGORY_COUNT> temp;
-        for (unsigned int err_idx = 0; err_idx < ZES_MAX_RAS_ERROR_CATEGORY_COUNT; err_idx++)
-            temp[err_idx] = pState.category[err_idx];
-        return temp;
+	return pState.category.at(errorcat);
```

Comment:
This looks like we fully populate the `pState.category` array then return just a single element for the requested category. Does that cancel any benefits we get from combining RAS queries in GEOPM batch reads?

Reply:
Note that the full array - `pState.category` is populated by L0 sysman, not PlatformIO. Each time a single element needs to be accessed, the entire array needs to be referenced. This overhead of populating all of contents of the `category` array is embedded within the execution model of L0. In fact, this latest commit removes the additional 'for' loop by removing the extra overhead of caching into 'temp' (on top of the aforementioned L0 internal overhead).